alf/utils/tensorrt_utils.py

```diff
@@ -0,0 +1,390 @@
+            alf.nest.flatten([args, kwargs]))
+        flat_all_args_np = [x.detach().cpu().numpy() for x in flat_all_args]
+        # engine accepts a list of args, not a tuple!
+        outputs_np = self._engine.run(flat_all_args_np)
+        # torch.from_numpy shares the memory with the numpy array
+        outputs = alf.nest.map_structure(
+            lambda x: torch.from_numpy(x).to(alf.get_default_device()),
+            outputs_np)
+        return self._onnx_wrapper.recover_module_output(outputs)
+
+
+def tensorrt_for_inference_if(cond: bool = True):
+    """A decorator to compile a method as a tensorRT engine for inference,
+    when ``cond`` is true.
+
+    This decorator should be used on a torch.nn.Module member function.
+
+    If will first wrap the object's method to export to onnx, and then create a
+    tensorrt engine from the onnx model. The engine will be cached in the object.
+    """
+
+    def _decorator(method):
+
+        if not cond:
+            return method
+
+        @functools.wraps(method)
+        def wrapped(module_to_wrap, *args, **kwargs):
+
+            # The first argument to the method is going to be ``self``, i.e. the
+            # instance that the method belongs to. By accessing it we get the
+            # reference of the module to wrap.
+            assert isinstance(module_to_wrap, torch.nn.Module), (
+                f'Cannot apply @tensorrt_for_inference_if on {type(module_to_wrap)}'
+            )
+
+            if not hasattr(module_to_wrap, '_tensorrt_engine_map'):
+                module_to_wrap._tensorrt_engine_map = {}
+            n_args = len(args)
+            arg_keys = tuple(kwargs.keys())
+            engine_key = (id(method), n_args, arg_keys)
+
+            tensorrt_engine = module_to_wrap._tensorrt_engine_map.get(
+                engine_key, None)
+            if tensorrt_engine is None:
+                tensorrt_engine = TensorRTEngine(
+                    module_to_wrap,
+                    method,
+                    example_args=args,
+                    example_kwargs=kwargs)
+                module_to_wrap._tensorrt_engine_map[
+                    engine_key] = tensorrt_engine
+                alf.utils.common.info(
+                    "Created a new TensorRT inference engine for "
+                    f"'{module_to_wrap.__class__.__name__}.{method.__name__}' "
+                    f"with key '{engine_key}'")
+            return tensorrt_engine(*args, **kwargs)
+
+        return wrapped
+
+    return _decorator
+
+
+_tensorrtified_methods = {}
+
+
+def tensorrtify_method(module, method_name):
+    """Convert a module method to use tensorrt inference at runtime. For example,
+
+    .. code-block:: python
+
+        if deploy_mode:
+            tensorrtify_method(agent, 'predict_step')
+
+        agent.predict_step(...)  # slow: prepare tensorrt engine for the first time
+        agent.predict_step(...)  # fast inference after the first call
+
```

Comment:
Should document the known major requirements or limitation for the method. For example, limitation on uint8, etc.

Reply:
> Should document the known major requirements or limitation for the method. For example, limitation on uint8, etc.

Added a 'README.md' file under this directory for a list of restrictions.